lib/lambda_ethereum_consensus/state_transition/state_transition.ex

```diff
@@ -3,14 +3,120 @@ defmodule LambdaEthereumConsensus.StateTransition do
   State transition logic.
   """
 
-  alias SszTypes.{BeaconState, SignedBeaconBlock}
+  alias LambdaEthereumConsensus.StateTransition
+  alias LambdaEthereumConsensus.StateTransition.EpochProcessing
+  alias SszTypes.{BeaconBlockHeader, BeaconState, SignedBeaconBlock}
 
+  @spec state_transition(BeaconState.t(), SignedBeaconBlock.t(), boolean()) ::
+          {:ok, BeaconState.t()} | {:error, String.t()}
   def state_transition(
         %BeaconState{} = state,
-        %SignedBeaconBlock{message: _block} = _signed_block,
+        %SignedBeaconBlock{message: block} = signed_block,
         _validate_result
       ) do
-    # TODO: implement
+    # NOTE: we aren't in a state to make validations yet
+    validate_result = false
+
+    state
+    # Process slots (including those with no blocks) since block
+    |> process_slots(block.slot)
+    # Verify signature
+    |> if_then_update(validate_result, fn state ->
+      if verify_block_signature(state, signed_block) do
+        {:ok, state}
+      else
+        {:error, "invalid block signature"}
+      end
+    end)
+    # Process block
+    |> map(&process_block(&1, block))
+    # Verify state root
+    |> map(
+      &if_then_update(&1, validate_result, fn state ->
+        if block.state_root != Ssz.hash_tree_root(state) do
```

Comment:
`hash_tree_root` returns {:ok, root}, right?

Reply:
Yes. I'll add a `hash_tree_root!` function for this, since the error can only happen because of API misuse (i.e. passing a non-`SszTypes` struct)